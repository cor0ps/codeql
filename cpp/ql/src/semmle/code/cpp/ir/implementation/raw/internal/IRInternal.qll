import semmle.code.cpp.ir.internal.IRCppLanguage as Language
import IRConstruction as Construction
