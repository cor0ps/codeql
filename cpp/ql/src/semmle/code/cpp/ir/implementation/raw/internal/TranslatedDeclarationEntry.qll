private import cpp
private import semmle.code.cpp.ir.implementation.Opcode
private import semmle.code.cpp.ir.implementation.internal.OperandTag
private import semmle.code.cpp.ir.internal.CppType
private import semmle.code.cpp.ir.internal.IRUtilities
private import InstructionTag
private import TranslatedElement
private import TranslatedExpr
private import TranslatedInitialization

/**
 * Gets the `TranslatedDeclarationEntry` that represents the declaration
 * `entry`.
 */
TranslatedDeclarationEntry getTranslatedDeclarationEntry(DeclarationEntry entry) {
  result.getAST() = entry
}

/**
 * Represents the IR translation of a declaration within the body of a function.
 * Most often, this is the declaration of an automatic local variable, although
 * it can also be the declaration of a static local variable. Declarations of extern variables and
 * functions do not have a `TranslatedDeclarationEntry`.
 */
abstract class TranslatedDeclarationEntry extends TranslatedElement, TTranslatedDeclarationEntry {
  DeclarationEntry entry;

  TranslatedDeclarationEntry() { this = TTranslatedDeclarationEntry(entry) }

  final override Function getFunction() {
    exists(DeclStmt stmt |
      stmt.getADeclarationEntry() = entry and
      result = stmt.getEnclosingFunction()
    )
  }

  final override string toString() { result = entry.toString() }

  final override Locatable getAST() { result = entry }
}

/**
 * Represents the IR translation of the declaration of a local variable,
 * including its initialization, if any.
 */
abstract class TranslatedLocalVariableDeclaration extends TranslatedVariableInitialization {
  /**
   * Gets the local variable being declared.
   */
  abstract LocalVariable getVariable();

  final override Type getTargetType() { result = getVariableType(getVariable()) }

  final override TranslatedInitialization getInitialization() {
    result =
      getTranslatedInitialization(getVariable().getInitializer().getExpr().getFullyConverted())
  }

  final override Instruction getInitializationSuccessor() {
    result = getParent().getChildSuccessor(this)
  }

  final override IRVariable getIRVariable() {
    result = getIRUserVariable(getFunction(), getVariable())
  }
}

/**
 * Represents the IR translation of a local variable declaration within a declaration statement.
 */
class TranslatedVariableDeclarationEntry extends TranslatedLocalVariableDeclaration,
  TranslatedDeclarationEntry {
  LocalVariable var;

  TranslatedVariableDeclarationEntry() { var = entry.getDeclaration() }

  override LocalVariable getVariable() { result = var }
}

/**
 * Gets the `TranslatedRangeBasedForVariableDeclaration` that represents the declaration of
 * `var`.
 */
TranslatedRangeBasedForVariableDeclaration getTranslatedRangeBasedForVariableDeclaration(
  LocalVariable var
) {
  result.getVariable() = var
}

/**
 * Represents the IR translation of a compiler-generated variable in a range-based `for` loop.
 */
class TranslatedRangeBasedForVariableDeclaration extends TranslatedLocalVariableDeclaration,
  TTranslatedRangeBasedForVariableDeclaration {
  RangeBasedForStmt forStmt;
  LocalVariable var;

  TranslatedRangeBasedForVariableDeclaration() {
    this = TTranslatedRangeBasedForVariableDeclaration(forStmt, var)
  }

  override string toString() { result = var.toString() }

  override Locatable getAST() { result = var }

  override Function getFunction() { result = forStmt.getEnclosingFunction() }

  override LocalVariable getVariable() { result = var }
}

TranslatedConditionDecl getTranslatedConditionDecl(ConditionDeclExpr expr) {
  result.getAST() = expr
}

/**
 * Represents the IR translation of the declaration portion of a `ConditionDeclExpr`, which
 * represents the variable declared in code such as:
 * ```
 * if (int* p = &x) {
 * }
 * ```
 */
class TranslatedConditionDecl extends TranslatedLocalVariableDeclaration, TTranslatedConditionDecl {
  ConditionDeclExpr conditionDeclExpr;

  TranslatedConditionDecl() { this = TTranslatedConditionDecl(conditionDeclExpr) }

  override string toString() { result = "decl: " + conditionDeclExpr.toString() }

  override Locatable getAST() { result = conditionDeclExpr }

  override Function getFunction() { result = conditionDeclExpr.getEnclosingFunction() }

  override LocalVariable getVariable() { result = conditionDeclExpr.getVariable() }
}
