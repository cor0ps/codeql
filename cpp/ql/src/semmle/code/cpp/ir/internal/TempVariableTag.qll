newtype TTempVariableTag =
  ConditionValueTempVar() or
  ReturnValueTempVar() or
  ThrowTempVar() or
  LambdaTempVar()

string getTempVariableTagId(TTempVariableTag tag) {
  tag = ConditionValueTempVar() and result = "CondVal"
  or
  tag = ReturnValueTempVar() and result = "Ret"
  or
  tag = ThrowTempVar() and result = "Throw"
  or
  tag = LambdaTempVar() and result = "Lambda"
}
