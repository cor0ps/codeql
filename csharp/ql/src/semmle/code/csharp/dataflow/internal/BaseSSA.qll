import csharp

/**
 * INTERNAL: Do not use.
 *
 * Provides a simple SSA implementation for local scope variables.
 */
module BaseSsa {
  private import ControlFlow
  private import AssignableDefinitions

  pragma[noinline]
  Callable getAnAssigningCallable(LocalScopeVariable v) {
    result = any(AssignableDefinition def | def.getTarget() = v).getEnclosingCallable()
  }

  private class SimpleLocalScopeVariable extends LocalScopeVariable {
    SimpleLocalScopeVariable() { not getAnAssigningCallable(this) != getAnAssigningCallable(this) }
  }

  /**
   * Holds if the `i`th node of basic block `bb` is assignable definition `def`,
   * targeting local scope variable `v`.
   */
  private predicate defAt(BasicBlock bb, int i, AssignableDefinition def, SimpleLocalScopeVariable v) {
    bb.getNode(i) = def.getAControlFlowNode() and
    v = def.getTarget() and
    // In cases like `(x, x) = (0, 1)`, we discard the first (dead) definition of `x`
    not exists(TupleAssignmentDefinition first, TupleAssignmentDefinition second | first = def |
      second.getAssignment() = first.getAssignment() and
      second.getEvaluationOrder() > first.getEvaluationOrder() and
      second.getTarget() = v
    )
  }

  /**
   * Holds if basic block `bb` would need to start with a phi node for local scope
   * variable `v` in an SSA representation.
   */
  private predicate needsPhiNode(BasicBlock bb, SimpleLocalScopeVariable v) {
    exists(BasicBlock def | def.inDominanceFrontier(bb) |
      defAt(def, _, _, v) or
      needsPhiNode(def, v)
    )
  }

  private newtype SsaRefKind =
    SsaRead() or
    SsaDef()

  /**
   * Holds if the `i`th node of basic block `bb` is a reference to `v`,
   * either a read (when `k` is `SsaRead()`) or a write including phi nodes
   * (when `k` is `SsaDef()`).
   */
  private predicate ssaRef(BasicBlock bb, int i, SimpleLocalScopeVariable v, SsaRefKind k) {
    bb.getNode(i).getElement() = v.getAnAccess().(VariableRead) and
    k = SsaRead()
    or
    defAt(bb, i, _, v) and
    k = SsaDef()
    or
    needsPhiNode(bb, v) and
    i = -1 and
    k = SsaDef()
  }

  /**
   * Gets the (1-based) rank of the reference to `v` at the `i`th node of basic
   * block `bb`, which has the given reference kind `k`.
   */
  private int ssaRefRank(BasicBlock bb, int i, SimpleLocalScopeVariable v, SsaRefKind k) {
    i = rank[result](int j | ssaRef(bb, j, v, _)) and
    ssaRef(bb, i, v, k)
  }

  /**
   * Holds if definition `def` of local scope variable `v` inside basic block
   * `bb` reaches the reference at rank `rnk`, without passing through another
   * definition of `v`, including phi nodes.
   */
  private predicate defReachesRank(
    BasicBlock bb, AssignableDefinition def, SimpleLocalScopeVariable v, int rnk
  ) {
    exists(int i | rnk = ssaRefRank(bb, i, v, SsaDef()) | defAt(bb, i, def, v))
    or
    defReachesRank(bb, def, v, rnk - 1) and
    rnk = ssaRefRank(bb, _, v, SsaRead())
  }

  /**
   * Holds if definition `def` of local scope variable `v` reaches the end of
   * basic block `bb` without passing through another definition of `v`, including
   * phi nodes.
   */
  private predicate reachesEndOf(AssignableDefinition def, SimpleLocalScopeVariable v, BasicBlock bb) {
    exists(int rnk |
      defReachesRank(bb, def, v, rnk) and
      rnk = max(ssaRefRank(bb, _, v, _))
    )
    or
    exists(BasicBlock mid |
      reachesEndOf(def, v, mid) and
      not exists(ssaRefRank(mid, _, v, SsaDef())) and
      bb = mid.getASuccessor()
    )
  }

  /**
   * Gets a read of the SSA definition for variable `v` at definition `def`. That is,
   * a read that is guaranteed to read the value assigned at definition `def`.
   */
  cached
  AssignableRead getARead(AssignableDefinition def, SimpleLocalScopeVariable v) {
    exists(BasicBlock bb, int i, int rnk |
      result.getAControlFlowNode() = bb.getNode(i) and
      rnk = ssaRefRank(bb, i, v, SsaRead())
    |
      defReachesRank(bb, def, v, rnk)
      or
      reachesEndOf(def, v, bb.getAPredecessor()) and
      not ssaRefRank(bb, _, v, SsaDef()) < rnk
    )
  }
}
