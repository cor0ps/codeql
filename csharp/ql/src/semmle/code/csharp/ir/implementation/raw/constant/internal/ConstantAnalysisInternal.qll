import semmle.code.csharp.ir.implementation.raw.IR as IR
