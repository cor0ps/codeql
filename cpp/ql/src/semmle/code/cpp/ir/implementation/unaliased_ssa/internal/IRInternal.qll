import semmle.code.cpp.ir.internal.IRCppLanguage as Language
import SSAConstruction as Construction
