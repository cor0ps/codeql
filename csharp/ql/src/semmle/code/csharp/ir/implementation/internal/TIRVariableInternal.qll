import semmle.code.csharp.ir.internal.IRCSharpLanguage as Language
import semmle.code.csharp.ir.implementation.raw.internal.IRConstruction as Construction
private import semmle.code.csharp.ir.implementation.TempVariableTag as TempVariableTag_

module Imports {
  module TempVariableTag = TempVariableTag_;
}
