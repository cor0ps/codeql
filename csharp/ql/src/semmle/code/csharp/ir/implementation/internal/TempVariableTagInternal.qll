import semmle.code.csharp.ir.internal.IRCSharpLanguage as Language
private import semmle.code.csharp.ir.internal.TempVariableTag as TempVariableTag_

module Imports {
  module TempVariableTag = TempVariableTag_;
}
