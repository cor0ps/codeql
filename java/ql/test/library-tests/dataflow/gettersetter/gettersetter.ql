import java
import semmle.code.java.dataflow.internal.DataFlowImplCommon
import semmle.code.java.dataflow.internal.DataFlowImplSpecific::Public
import semmle.code.java.dataflow.internal.DataFlowImplSpecific::Private

private predicate read(Node n1, Content f, Node n2) {
  readDirect(n1, f, n2) or
  argumentValueFlowsThrough(_, n1, TContentSome(f), TContentNone(), n2)
}

private predicate store(Node n1, Content f, Node n2) {
  storeDirect(n1, f, n2) or
  argumentValueFlowsThrough(_, n1, TContentNone(), TContentSome(f), n2)
}

from Node n1, Content f, Node n2, string k
where
  read(n1, f, n2) and k = "Read"
  or
  store(n1, f, n2) and k = "Store"
select k, n1, n2, f
