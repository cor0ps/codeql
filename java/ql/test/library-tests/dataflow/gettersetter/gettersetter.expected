| Read | A.java:5:12:5:15 | this | A.java:5:12:5:19 | this.foo | A.java:2:7:2:9 | foo |
| Read | A.java:21:13:21:13 | a | A.java:21:13:21:22 | getFoo(...) | A.java:2:7:2:9 | foo |
| Read | A.java:23:9:23:9 | a | A.java:23:9:23:19 | aGetter(...) | A.java:2:7:2:9 | foo |
| Read | A.java:24:9:24:10 | a2 | A.java:24:9:24:23 | notAGetter(...) | A.java:2:7:2:9 | foo |
| Read | A.java:45:12:45:38 | maybeIdWrap(...) | A.java:45:12:45:42 | maybeIdWrap(...).foo | A.java:2:7:2:9 | foo |
| Read | A.java:49:12:49:38 | maybeIdWrap(...) | A.java:49:12:49:42 | maybeIdWrap(...).foo | A.java:2:7:2:9 | foo |
| Store | A.java:9:16:9:16 | x | A.java:9:5:9:8 | this [post update] | A.java:2:7:2:9 | foo |
| Store | A.java:14:13:14:13 | x | A.java:14:5:14:5 | a [post update] | A.java:2:7:2:9 | foo |
| Store | A.java:20:14:20:14 | 1 | A.java:20:5:20:5 | a [post update] | A.java:2:7:2:9 | foo |
| Store | A.java:22:20:22:20 | 2 | A.java:22:12:22:21 | withFoo(...) | A.java:2:7:2:9 | foo |
